### ISSUE TITLE
Cannot log in Response code 400 (Bad Request)

### ISSUE THREAD

**Issue by decafbad86** [Hacksore/bluelinky#89]
Hi,

today I tried the bluelinky lib on two systems - MacOS 10.13 and windows 10 - both got the same node.js environment. As shown in docs I changed my credentials an my VIN. When now trying to log in / creating the client as shown I always get 

Logging in...
(node:68683) UnhandledPromiseRejectionWarning: Error [ERR_UNHANDLED_ERROR]: Unhandled error. ('Response code 400 (Bad Request)')
    at t.emit (events.js:304:17)
    at t.<anonymous> (/Users/thomas/node_modules/bluelinky/dist/index.js:2:41053)
    at /Users/thomas/node_modules/bluelinky/dist/index.js:2:1955
    at Object.throw (/Users/thomas/node_modules/bluelinky/dist/index.js:2:2060)
    at a (/Users/thomas/node_modules/bluelinky/dist/index.js:2:850)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

The region is EU - I can log in via web, I can log in via iOS App. 
The console output is from debug.js.
The car it self has the new infotainment software - no three splitted screen.

I tried to downgrade from 5.2.3 to 5.2.2

**Comment by PierreLevres** [comment]
Logout from app on phone. Logon again. You might have to change pass or pin


Groet,
Peter Lips

**Comment by decafbad86** [comment]
Thanks for the quick reply - I tried this a two time - and changed the password manually for this - this was mentioned in an issue erlier. The moment you wrote - I again logged out and in again on the phone. There was no notification to change the PIN nor the password.

Does it matter if i use the Hyundai account or the old blue link account?

**Comment by dpembo** [comment]
I get same issue having gone through the hyundai account changes .  Will try and dig a little.

**Comment by dpembo** [comment]
It fails on the auth unsurprisingly.

I think @decafbad86 you were right.

For me - my original bluelink account and my hyundai account have different email addresses.
If you're logged into the blue link app, take a look at your account and see what email address is in use.  This now shows the hyundai account for me, not the original bluelink account.  

If I login using the hyundai account, it seems to work.

```
...
[2021-01-17 09:00:02] debug: Getting cars from controller
[2021-01-17 09:00:03] debug: EU Vehicle xxxxxxxxx created
[2021-01-17 09:00:03] debug: Added vehicle xxxxxxxxx
[2021-01-17 09:00:03] debug: Found 1 on the account
...
```

**Comment by Hacksore** [comment]
So for my understanding, the EU accounts went through some kind of migration, right?

**Comment by dpembo** [comment]
@Hacksore, kind of - it looks like they're trying to standardise on one account for all hyundai services so are asking you to move to the hyundai account (and create if necessary) versus using the original account registered against bluelink.

To be honest, the process was a little confusing as the login screen on the app shows the below right now.
![bluelink](https://user-images.githubusercontent.com/9807493/104857550-2ed59f00-5911-11eb-8a69-3915e326272a.jpg)

If you login with the original account (using current account login button), It asks you to upgrade your account to the hyundai account, and thereafter you have to use this to login approach going forward both in the app (through the hyundai account login button) which takes you to another login screen, and this account also works against the APIs.

**Comment by mpucherSmIT** [comment]
Hi, still no luck for me.

I upgraded my bluelink account to an hyundai account (created a new one with the same e-mail). Changed PIN, password and upgraded to the latest bluelinky npm package (5.2.3). Logged out from the app and in a few times. Still stuck at "logging in to API..."

What else could I do?

Thank you!

```
mike@mikenet:~/projects/energylog/bluelink$ LOG_LEVEL=debug node kona.js 
[2021-01-18 15:23:33] debug: EU Controller created
[2021-01-18 15:23:33] debug: Bluelinky is logging in automatically, to disable use autoLogin: false
[2021-01-18 15:23:33] debug: loggin into to API
(node:25754) UnhandledPromiseRejectionWarning: Error [ERR_UNHANDLED_ERROR]: Unhandled error. ('Response code 400 (Bad Request)')
    at t.emit (events.js:303:17)
    at t.<anonymous> (/home/mike/projects/energylog/bluelink/node_modules/bluelinky/dist/index.js:2:41136)
    ....
```

From my config:


```
const BlueLinky = require('bluelinky');

const client = new BlueLinky({
  username: 'john@gmail.com',
  password: 'new_hyundai_account_pw',
  region: 'EU',
  pin: 'new_pin_changed_in_app_with_hyundai_account'
});
```

**Comment by dpembo** [comment]
Hmmm... the one thing I did do was keep the same password (it's some randomly keepass generated thing) so the hyundai account and my bluelink account actually shared the same password but with different email addresses.

Do you see the same login screen I screen grabbed if you logout of bluelink and log back in?

If you try to login with the current account, does it show your account as being 'upgraded' and therefore you need to login with the hyundai acccount?

I was testing with the develop branch of bluelinky (after fixing it up a little), but I don't believe the changes in there would make so much difference.  I use node-red (not the bluelinky library) to do the actual integration calls as this is how I originally put it together, using the excellent bluelinky to save me having to rip apart the android application to understand the calls.  Once I'd sorted the account, my node-red implementation also started to work again with no changes other than the account username, so the actual implementation hasn't really changed at all, and all API calls remain the same.

**Comment by mpucherSmIT** [comment]
> Do you see the same login screen I screen grabbed if you logout of bluelink and log back in?

Yes.
 
> If you try to login with the current account, does it show your account as being 'upgraded' and therefore you need to login with the hyundai acccount?

Exactly. It shows "upgrade completed" and redirects to the "Hyundai Account Login" Screen.

**Comment by mpucherSmIT** [comment]
The only difference is that the Hyundai account forces you to use special characters in the password. Could that lead to problems in the password string and during the login call?

**Comment by decafbad86** [comment]
If it helps - I do have an Hyundai account and I traced the traffic with SSL decrypt with Chales Proxy. I cleaned some data from this file - my phone number and fullname
[BlueLink.chlsx.zip](https://github.com/Hacksore/bluelinky/files/5831329/BlueLink.chlsx.zip)
The actions I sniffed were logging in and getting the status

**Comment by dpembo** [comment]
> If it helps - I do have an Hyundai account and I traced the traffic with SSL decrypt with Chales Proxy. I cleaned some data from this file - my phone number and fullname
> [BlueLink.chlsx.zip](https://github.com/Hacksore/bluelinky/files/5831329/BlueLink.chlsx.zip)
> The actions I sniffed were logging in and getting the status

I've been trying to bypass the certificate pinning on this for a while unsuccessfully so I can trace the calls 😞. 
I'm trying to figure out why a few of the API calls don't totally represent what the app does for an EV!

If you can tell me how you manage to do this, that would be great! I tried objection, but that app just crashed out having patched it...

**Comment by decafbad86** [comment]
Nobody expected this to be the full representation of the BlueLink API - this is just login and a get status call - the basic things to get things working. 

Okay - now to patching - it is not possible to patch the phone - in my case an iPhone - patching will open up a VPN and Hyundai does not like any VPN - so - in my environment I 

1. set up my old MacBook installed Chales Web Proxy (https://www.charlesproxy.com/) from here 
2. followed the docs on their page to set up a proxy in iOS for wireless lan
3. followed the docs to get a SSL man in the middle certificate to install this in iOS - be sure using Safari at this point as default browser
4. confirmed the use of this certificate - also written down in their docs
5. go to Mac and start sniffing - to be sure to get only the Hyundai calls - clean the list of web responses and requests on the left
6. You are done - "just analyse it"

Maybe with Android and Windows Fiddler will do this job for you - but I can't say anything about that.

**Comment by dpembo** [comment]
> Nobody expected this to be the full representation of the BlueLink API - this is just login and a get status call - the basic things to get things working.
> 
> Okay - now to patching - it is not possible to patch the phone - in my case an iPhone - patching will open up a VPN and Hyundai does not like any VPN - so - in my environment I
> 
> 1. set up my old MacBook installed Chales Web Proxy (https://www.charlesproxy.com/) from here
> 2. followed the docs on their page to set up a proxy in iOS for wireless lan
> 3. followed the docs to get a SSL man in the middle certificate to install this in iOS - be sure using Safari at this point as default browser
> 4. confirmed the use of this certificate - also written down in their docs
> 5. go to Mac and start sniffing - to be sure to get only the Hyundai calls - clean the list of web responses and requests on the left
> 6. You are done - "just analyse it"
> 
> Maybe with Android and Windows Fiddler will do this job for you - but I can't say anything about that.

thanks - I'll try it on an apple device - maybe this is less secure.

The android version stops mitm sniffing by certificate pinning.  In essence, it ensures the cert it gets from the destination matches what it expects inside the app code, so the only way to sniff the traffic is to 'modify' the app to remove this check.  On android this means modifying the bytecode to remove the check, and it's not for the feint hearted despite knowing java well 😄

**Comment by decafbad86** [comment]
Let me know if I can help you - in the evening I will have access again to my Mac

**Comment by dpembo** [comment]
I had a quick look in your log... I can't see this call anywhere in your log.
https://prd.eu-ccapi.hyundai.com:8080/api/v1/user/signin

Which I'm pretty certain is where it was failing for me.  The OAuth flow looks slightly different also, but I'm not sure if this is an ios/android difference I'm seeing.

Remove your proxy log btw - I'm pretty certain there's possibly things in there hiding in base64 you don't really want to share.

**Comment by decafbad86** [comment]
Thanks for this hint - I'm pretty sure that I know what you mean.... uhm as I can remember - I did not track the web based login - just starting the app - as I walked this through yesterday evening - there was a missing link in my head where the f... the app is transmitting the basic login....

**Comment by mpucherSmIT** [comment]
Weird behavior: after playing around with debug.js I figured out that I can log in to the API with the old bluelink account pw and the newly assigned PIN (from the hyundai account).

So far my script works but strange things are prob. going on on hyundai side...

**Comment by Hacksore** [comment]
FWIW if you have a rooted Android phone you can use this method to get
around both root detection and SSL pinning.

https://docs.google.com/document/d/1q0Pz2t5Gmt-vv8z-2Yswhvg2x9B2f3d1vjpkIjt9EfQ/edit

**Comment by dpembo** [comment]
So the apple version clearly doesn't worry about certificate pinning so it seems.

Looking at the login flow, it now goes to `https://eu-account.hyundai.com/` to do some authentication

`https://eu-account.hyundai.com/auth/realms/euhyundaiidm/login-actions/authenticate?session_code=<session_id>&execution=<execid>&client_id=<client id>`

Posting the data

`username=<email>&password=<password>&rememberMe=on&credentialId=
`
This then bounces around quite a lot through http302 responses returning different locations and cookies. (Seems to use keycloak).

1st 302 = Location Header
`https://eu-account.hyundai.com/auth/realms/euhyundaiidm/login-actions/required-action?execution=kia-account-upgrade-email&client_id=<clientid>&tab_id=<tabid>`

2nd 302 = Location Header
`https://prd.eu-ccapi.hyundai.com:8080/api/v1/user/integration/redirect/login?user_id=4xxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxc&state=<stateguid>&session_state=<sessionguid>&code=<codeguid>`

3rd 302 =Location Header
`https://prd.eu-ccapi.hyundai.com:8080/web/v1/user/integration/auth?userId=axxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`

Eventually this then posts to:
`POST https://prd.eu-ccapi.hyundai.com:8080/api/v1/user/silentsignin`

with the body containing the second userid guid
`{"userId":"axxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"}`


This responds with a 
```
{
  "redirectUrl": "https://prd.eu-ccapi.hyundai.com:8080/api/v1/user/oauth2/redirect?code=<codeID>&state=test",
  "popup": false,
  "method": "",
  "upgrade": false,
  "integrated": false,
  "deleteAccountLink": ""
}

```
then it goes off into the rest of the auto process following the redirect.

**Comment by Hacksore** [comment]
@dpembo after you upgrade your account do you essentially break yourself from using the library now?

**Comment by dpembo** [comment]
No, right now it still currently works somehow!
I kept the same password, and the same pin, but a different email address and I can login through bluelinky still, using the newer email address, so whatever has changed behind the scene has likely modified the bluelink account.

If I try to login with my old bluelink email address in the app, it tells me that my account has been upgraded, and I have to re-login using the hyundai login.

**Comment by Hacksore** [comment]
@dpembo Would it make sense to just inform users they must migrate their account in order to ensure using bluelinky in EU will work.

If there is nothing really to change from a code perspective that'd be my guidance but then again I'm not part of EU so I'm a bit biased 😀.

**Comment by dpembo** [comment]
Yes, I had to migrate my account, but I kept the same password, and the same pin (had a different email address).  I can login through bluelinky still, using the newer email address.  If I try to login  the app with my old bluelink email address, it tells me that my account has been upgraded, and I have to re-login using the hyundai login, so something behind the scenes must have created a relation between these accounts.

When I get more time, I'll step piece by piece through the login again and update my integrations (had to steal my daughters iPad to trace the flow 😄), then I'll use the knowledge to update bluelinky EU and create a pull request on the develop branch.

For now it works for both me and @mpucherSmIT so I'd suggest you inform EU users that they might need to migrate their account and close this for now.  It may come up again, either when US or Canada login process changes or they disable the old login mechanism in the EU - hopefully before then I'll have sorted the login code
